HIRS_ProvisionerTPM2/src/tpm_version.cpp

```diff
@@ -1,11 +1,190 @@
-    cout << "TPM Vendor ID: UNKN" << endl;
+    TPMS_CAPABILITY_DATA capability_data;
+    TPMI_YES_NO            more_data;
+    shared_ptr<SapiContext> sapiContext = SapiContext::create();
+    TSS2_RC rc;
+    do {
+        // The return code will be 4 bytes. The most significant byte
+        // tells what software layer is reposible for a non-successful
+        // attempt at executing the call. The other bytes provide the error
+        // type. If we get a retry warning, we don't care which layer is
+        // responsible, so we mask the lower 3 bytes and compare it to
+        // TSS2_RC_RETRY to see if we need to try again.
+        rc = Tss2_Sys_GetCapability(sapiContext->getPointer(), nullptr,
+                               TPM_CAP_TPM_PROPERTIES,
+                               PT_FIXED,
+                               MAX_TPM_PROPERTIES,
+                               &more_data,
+                               &capability_data,
+                               nullptr);
+    } while ((rc & 0xfff) == TPM_RC_RETRY);
+
+    // There is nothing we can do if an error occurred. The HIRS Client will
+    // know how to respond if this happens. No need to log anything.
+    if (rc != TSS2_RC_SUCCESS) {
+        return 1;
+    }
+
+    // We only need to collect the following data from the TPM.
+    string manufacturer;
+    string majorVersion;
+    float minorVersion;
+
+    for (size_t i = 0; i < capability_data.data.tpmProperties.count; ++i) {
+        TPMS_TAGGED_PROPERTY & p
+                = capability_data.data.tpmProperties.tpmProperty[i];
+        TPM_PT property = p.property;
+
+        // All data is in the form of a UINT32, even if it represents a string.
+        // For strings, the bytes are in the wrong endianness and are not
+        // null-terminated.
+        UINT32 value = p.value;
+
+        switch (property) {
+            case TPM_PT_FAMILY_INDICATOR:
+                majorVersion = reversedStringValue(value);
+                break;
+            case TPM_PT_REVISION:
+                // The minor version has two decimal places, but since it is
+                // stored as an integer, it is stored as 100 times its value.
+                minorVersion = value / 100.0f;
+                break;
+            case TPM_PT_MANUFACTURER:
+                manufacturer = reversedStringValue(value);
+                break;
+        }
+    }
+
+    cout << "Chip Version: " << majorVersion << "." << minorVersion << endl;
+    cout << "TPM Vendor ID: " << manufacturer << endl;
     return 0;
 }
 
+/**
+ * Returns the 4-byte string represented by the bytes of value.
+ *
+ * Assumes the bytes of value are reversed.
+ * @param value the 4 bytes to be reversed and placed in the returned string
+ * @return the string represented by value
+ */
+string reversedStringValue(UINT32 value) {
+    string stringValue(4, 0);  // make a string on length 4
+    std::memcpy(&stringValue[0], &value, sizeof value);  // copy bytes in
```

Comment:
Suggest referencing the size of the target versus the source when specifying size for the memcpy.

Reply:
okay.